pkg/compose/build_buildkit.go

```diff
@@ -20,23 +20,120 @@ import (
+	"strings"
+	"unicode"
 
 	_ "github.com/docker/buildx/driver/docker"           //nolint:blank-imports
 	_ "github.com/docker/buildx/driver/docker-container" //nolint:blank-imports
 	_ "github.com/docker/buildx/driver/kubernetes"       //nolint:blank-imports
 	_ "github.com/docker/buildx/driver/remote"           //nolint:blank-imports
-	"github.com/moby/buildkit/client"
+	"github.com/docker/compose/v2/pkg/progress"
+	bclient "github.com/moby/buildkit/client"
+	"github.com/opencontainers/go-digest"
 
 	"github.com/docker/buildx/build"
 	"github.com/docker/buildx/builder"
 	"github.com/docker/buildx/util/dockerutil"
 	xprogress "github.com/docker/buildx/util/progress"
-	"github.com/docker/compose/v2/pkg/progress"
 )
 
-func (s *composeService) doBuildBuildkit(ctx context.Context, opts map[string]build.Options, mode string) (map[string]string, error) {
+type BuildOutput struct {
+	service  string
+	vertexes map[digest.Digest]*bclient.Vertex
+	logs     map[digest.Digest]int
+	w        progress.Writer
+}
+
+func sanitize(s string) string {
+	split := strings.Split(strings.TrimSpace(s), "\n")
+	return split[len(split)-1]
+}
+
+func (b *BuildOutput) Write(status *bclient.SolveStatus) {
+	for _, v := range status.Vertexes {
+		b.vertexes[v.Digest] = v
+
+		txt := v.Name
+		if v.Cached {
+			txt = "CACHED " + v.Name
+		}
+		if v.ProgressGroup != nil {
+			txt = txt + " => " + v.ProgressGroup.String()
+		}
+		status := progress.Working
+		if v.Completed != nil {
+			status = progress.Done
+		}
+
+		b.w.Event(progress.Event{
+			ParentID:   b.service,
+			ID:         v.Digest.String(),
+			StatusText: sanitize(txt),
+			Status:     status,
+		})
+
+	}
+
+	for _, s := range status.Statuses {
+		txt := s.Name
+		if s.Current > 0 {
+			txt = fmt.Sprintf("%s %dB", s.Name, s.Current)
+		}
+		b.w.Event(progress.Event{
+			ParentID:   b.service,
+			ID:         b.getVertexId(s.Vertex),
+			StatusText: sanitize(txt),
+			Status:     progress.Done,
+		})
+	}
+	for _, s := range status.Warnings {
+		b.w.Event(progress.Event{
+			ParentID:   b.service,
+			ID:         b.getVertexId(s.Vertex),
+			StatusText: sanitize(string(s.Short)),
+		})
+	}
+	for _, l := range status.Logs {
+		var start int
+		count := b.logs[l.Vertex]
+
+		for idx, r := range l.Data {
+			if r == '\n' || r == '\r' {
+				line := string(l.Data[start:idx])
+				start = idx + 1
+				if len(line) == 0 {
+					continue
+				}
+				line = strings.TrimRightFunc(line, unicode.IsSpace)
+				b.w.Event(progress.Event{
+					ParentID:   b.getVertexId(l.Vertex),
+					ID:         fmt.Sprintf("#%d", count),
+					StatusText: line,
+					Status:     progress.Log,
+				})
+				count++
+			}
+		}
+		b.logs[l.Vertex] = count
+	}
+}
+
+func (b *BuildOutput) ValidateLogSource(digest digest.Digest, i interface{}) bool {
+	return true
+}
+
+func (b BuildOutput) ClearLogSource(i interface{}) {
+
+}
+
+func (b *BuildOutput) getVertexId(vertex digest.Digest) string {
+	return b.vertexes[vertex].Digest.String()
+}
```

Comment:
Can we move all this part in a dedicated file such as `pkg/compose/buildkitoutput`?